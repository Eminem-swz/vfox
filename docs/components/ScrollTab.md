# ScrollTab 标签滚动布局

## ScrollTab Props

| 属性                 | 类型          | 默认值 | 必填 | 说明                        |
| -------------------- | ------------- | ------ | ---- | --------------------------- |
| sticky-offset-top    | number/string | 0      | 否   | 数值默认是 px，也支持 vw/vh |
| sticky-offset-buttom | number/string | 0      | 否   | 数值默认是 px，也支持 vw/vh |

## ScrollTab Events

| 事件   | 描述       | 回调函数参数               |
| ------ | ---------- | -------------------------- |
| change | 切换时触发 | { activeIndex } 当前项索引 |

## ScrollTab Slots

注：其中只可放置 [ScrollTabItem](./ScrollTab.md#scrolltabitem-标签滚动子项) 组件，否则会导致未定义的行为。

```
<fx-scroll-tab>
  <fx-scroll-tab-item name="Dust Red">
    <div class="scroll-tab-box box-1"></div>
  </fx-scroll-tab-item>
  <fx-scroll-tab-item name="Volcano">
    <div class="scroll-tab-box box-2"></div>
  </fx-scroll-tab-item>
  <fx-scroll-tab-item name="Sunset Orange">
    <div class="scroll-tab-box box-3"></div>
  </fx-scroll-tab-item>
  ...
</fx-scroll-tab>
```

## ScrollTabItem 标签滚动子项

## ScrollTabItem Props

| 属性 | 类型   | 默认值 | 必填 | 说明                       |
| ---- | ------ | ------ | ---- | -------------------------- |
| name | string |        | 是   | 分组名，也应用于吸附和菜单 |

## ScrollTabItem Slots

```
<fx-scroll-tab-item name="Dust Red">
  <div class="scroll-tab-box box-1"></div>
</fx-scroll-tab-item>
```
