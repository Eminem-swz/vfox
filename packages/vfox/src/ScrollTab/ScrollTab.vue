<template>
  <div class="fx-scroll-tab">
    <div class="fx-scroll-tab_sidebar">
      <Sticky
        ref="side"
        :offsetTop="stickyOffsetTop"
        :offsetBottom="stickyOffsetBottom"
      >
        <SideTab :options="tabList" v-model:activeValue="activeIndex" />
      </Sticky>
    </div>
    <div class="fx-scroll-tab_body">
      <StickyView
        :offsetTop="stickyOffsetTop"
        ref="body"
        v-model:activeIndex="activeIndex"
        @resetItems="onResetItems"
        @change="onChange"
      >
        <slot></slot>
      </StickyView>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref } from 'vue'
import { SideTab } from '../SideTab'
import { Sticky } from '../Sticky'
import { StickyView } from '../StickyView'
import { sizeValidator } from '../helpers/validator'
import type { ScrollToIndexOptions, ScrollToOptions } from '../StickyView/types'
import type { OnChange as StickyViewOnChange } from '../StickyView/types'
import { emitChangeValidator } from '../StickyView/stickyView'
import type { DomSelector } from '../helpers/types'

export default defineComponent({
  name: 'fx-scroll-tab',
  components: { SideTab, Sticky, StickyView },
  props: {
    stickyOffsetTop: {
      validator: sizeValidator,
      default: 0
    },
    stickyOffsetBottom: {
      validator: sizeValidator,
      default: 0
    }
  },
  emits: {
    change: emitChangeValidator
  },
  setup(props, { emit }) {
    const side = ref<InstanceType<typeof Sticky>>()
    const body = ref<InstanceType<typeof StickyView>>()
    const tabList = ref<
      {
        value: number
        label: string
      }[]
    >([])
    const activeIndex = ref(0)

    function resetContainer(containSelector: DomSelector) {
      side.value?.resetContainer(containSelector)
      body.value?.resetContainer(containSelector)
    }

    function onResetItems(items: { name: string; index: number }[]) {
      tabList.value = items.map(item => {
        return {
          value: item.index,
          label: item.name
        }
      })
    }

    const onChange: StickyViewOnChange = res => {
      emit('change', res)
    }

    /**
     * 滚动到第index个
     * @param index 索引
     */
    function scrollToIndex(index: number | ScrollToIndexOptions) {
      body.value && body.value.scrollToIndex(index)
    }

    /**
     * 滚到到指定位置
     * @param scrollTop 位置值
     */
    function scrollTo(scrollTop: number | ScrollToOptions) {
      body.value && body.value.scrollTo(scrollTop)
    }

    onMounted(() => resetContainer(document))

    return {
      side,
      body,
      activeIndex,
      tabList,
      onChange,
      scrollToIndex,
      scrollTo,
      resetContainer,
      onResetItems
    }
  }
})
</script>
