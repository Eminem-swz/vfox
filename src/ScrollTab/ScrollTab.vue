<template>
  <div class="fx-scroll-tab">
    <div class="fx-scroll-tab_sidebar">
      <Sticky
        ref="sidebar"
        :offsetTop="stickyOffsetTop"
        :offsetBottom="stickyOffsetBottom"
      >
        <SideTab :options="tabList" v-model:activeValue="activeIndex" />
      </Sticky>
    </div>
    <div class="fx-scroll-tab_body">
      <StickyView
        :offsetTop="stickyOffsetTop"
        ref="body"
        v-model:activeIndex="activeIndex"
        @resetItems="onResetItems"
        @change="onChange"
      >
        <slot></slot>
      </StickyView>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref } from 'vue'
import { SideTab } from '@/SideTab'
import { Sticky } from '@/Sticky'
import { StickyView } from '@/StickyView'
import { sizeValidator } from '@/helpers/validator'
import type { ScrollToIndexOptions, ScrollToOptions } from '../helpers/types'
import type { StickyViewChangeArgs } from '@/StickyView/types'
import { emitChangeValidator } from '@/StickyView/stickyView'

export default defineComponent({
  name: 'fx-scroll-tab',
  components: { SideTab, Sticky, StickyView },
  props: {
    stickyOffsetTop: {
      validator: sizeValidator,
      default: 0
    },
    stickyOffsetBottom: {
      validator: sizeValidator,
      default: 0
    }
  },
  emits: {
    change: emitChangeValidator
  },
  setup(props, { emit }) {
    const sidebar = ref()
    const body = ref()
    const tabList = ref<
      {
        value: number
        label: string
      }[]
    >([])
    const activeIndex = ref(0)

    function resetContainer(containSelector: any) {
      sidebar.value && sidebar.value.resetContainer(containSelector)
      body.value && body.value.resetContainer(containSelector)
    }

    function onResetItems(items: { name: string; index: number }[]) {
      tabList.value = items.map(item => {
        return {
          value: item.index,
          label: item.name
        }
      })
    }

    function onChange(res: StickyViewChangeArgs) {
      emit('change', res)
    }

    /**
     * 滚动到第index个
     * @param index 索引
     */
    function scrollToIndex(index: number | ScrollToIndexOptions) {
      body.value && body.value.scrollToIndex(index)
    }

    /**
     * 滚到到指定位置
     * @param scrollTop 位置值
     */
    function scrollToOffset(scrollTop: number | ScrollToOptions) {
      body.value && body.value.scrollToOffset(scrollTop)
    }

    onMounted(() => resetContainer(document))

    return {
      sidebar,
      body,
      activeIndex,
      tabList,
      onChange,
      scrollToIndex,
      scrollToOffset,
      resetContainer,
      onResetItems
    }
  }
})
</script>
