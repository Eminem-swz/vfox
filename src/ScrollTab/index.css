.fx-scroll-tab {
  display: flex;
  overflow-y: auto;
  position: relative;
  background-color: #f5f5f5;
}

.fx-scroll-tab_sidebar {
  width: 80px;
  height: 100%;
  flex-grow: 0;
  background-color: #f5f5f5;
}

.fx-scroll-tab_body {
  flex: 1;
}
