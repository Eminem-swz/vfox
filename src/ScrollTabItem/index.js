import ScrollTabItem from '../Tab/ScrollTabItem.vue'

ScrollTabItem.install = function(app) {
  app.component(ScrollTabItem.name, ScrollTabItem)
}

export default ScrollTabItem
