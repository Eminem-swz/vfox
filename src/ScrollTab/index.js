import ScrollTab from '../Tab/ScrollTab.vue'

ScrollTab.install = function(app) {
  app.component(ScrollTab.name, ScrollTab)
}

export default ScrollTab
