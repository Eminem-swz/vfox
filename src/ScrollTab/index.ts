import { App } from 'vue'
import { SFCWithInstall } from '@/utils/types'
import ScrollTab from '../Tab/ScrollTab.vue'

const _ScrollTab: SFCWithInstall<typeof ScrollTab> = Object.assign(ScrollTab, {
  install: function(app: App) {
    app.component(ScrollTab.name, ScrollTab)
  }
})

export default _ScrollTab
