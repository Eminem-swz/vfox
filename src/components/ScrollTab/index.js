import ScrollTab from '../Tab/Scroll.vue'

ScrollTab.install = function(Vue) {
  Vue.component(ScrollTab.name, ScrollTab)
}

export default ScrollTab
