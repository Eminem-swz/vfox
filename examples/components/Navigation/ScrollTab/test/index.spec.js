/* global describe,test,expect */

import { mount } from '@vue/test-utils'
import ScrollTab from '@/ScrollTab'
import ScrollTabItem from '@/ScrollTabItem'

describe('ScrollTab', () => {
  test('ScrollTab should render default correctly', () => {
    const wrapper = mount(ScrollTab, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('ScrollTabItem should render default correctly', () => {
    const wrapper = mount(ScrollTabItem, {
      propsData: { name: 'item-1' },
      slots: {
        default: 'content'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
