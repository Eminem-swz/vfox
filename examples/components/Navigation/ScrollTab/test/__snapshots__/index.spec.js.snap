// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollTab ScrollTab should render default correctly 1`] = `
<div class="fx-scroll-tab">
  <div class="fx-scroll-tab_sidebar">
    <div class="fx-sticky">
      <div class="fx-sticky_content" style="top: 0px; left: 0px; width: 0px; bottom: 0px; z-index: 1000; position: fixed;">
        <div class="fx-side-tab">
          <ul class="fx-side-tab_list"></ul>
        </div>
      </div>
    </div>
  </div>
  <div class="fx-scroll-tab_body">
    <div class="fx-sticky-view">
      <div class="fx-sticky-view_list"></div>
      <div class="fx-sticky fx-sticky-view_top">
        <div class="fx-sticky_content" style="top: 0px; left: 0px; width: 0px; height: 0px; z-index: 1000; position: fixed;">
          <div class="fx-sticky-view_fixed" style="transform: translate3d(0px, 0px, 0px);"></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ScrollTab ScrollTabItem should render default correctly 1`] = `
<div class="fx-sticky-view-item fx-scroll-tab-item">
  <div class="fx-sticky-view-item_header">item-1</div>
  <div class="fx-sticky-view-item_body">content</div>
</div>
`;
